core/unit_test/TestMDRange.hpp

```diff
@@ -2057,6 +2074,54 @@ struct TestMDRange_4D {
       ASSERT_EQ(counter, 0);
     }
   }  // end test_for4
+
+  // test that each iteration is evaluated only once
+  // see #7697
+  static void test_for4_eval_once(const int N0, const int N1, const int N2,
+                                  const int N3) {
+    test_for4_eval_once_iterate<Iterate::Left, Iterate::Left>(N0, N1, N2, N3,
+                                                              "LL");
+    test_for4_eval_once_iterate<Iterate::Left, Iterate::Right>(N0, N1, N2, N3,
+                                                               "LR");
+    test_for4_eval_once_iterate<Iterate::Right, Iterate::Left>(N0, N1, N2, N3,
+                                                               "RL");
+    test_for4_eval_once_iterate<Iterate::Right, Iterate::Right>(N0, N1, N2, N3,
+                                                                "RR");
+  }  // end test_for4_eval_once
+
+ private:
+  // test that each iteration is evaluated only once for a specified iteration
+  // order
+  template <Iterate Outer, Iterate Inner>
+  static void test_for4_eval_once_iterate(const int N0, const int N1,
+                                          const int N2, const int N3,
+                                          const char *iterate_msg) {
+    using range_type =
+        typename Kokkos::MDRangePolicy<ExecSpace, Kokkos::Rank<4, Outer, Inner>,
+                                       Kokkos::IndexType<int>, AtomicTag>;
+    using point_type = typename range_type::point_type;
+
+    range_type range(point_type{{0, 0, 0, 0}}, point_type{{N0, N1, N2, N3}});
+
+    TestMDRange_4D functor(N0, N1, N2, N3);
+
+    parallel_for(range, functor);
+    MinMaxValue min_max_value;
+    DataType sum_value;
+    parallel_reduce(range, functor, MinMax(min_max_value), sum_value);
+
+    if (min_max_value.min_val != 1 or min_max_value.max_val != 1) {
+      printf(
+          " Errors in test_for4_eval_once for shape (%d, %d, %d, %d) and "
+          "iterate %s; min = %d, max = %d\n\n",
+          N0, N1, N2, N3, iterate_msg, min_max_value.min_val,
+          min_max_value.max_val);
+    }
+
+    ASSERT_EQ(min_max_value.min_val, 1);
+    ASSERT_EQ(min_max_value.max_val, 1);
+    ASSERT_EQ(sum_value, N0 * N1 * N2 * N3);
```

Comment:
```suggestion
    ASSERT_EQ(min_max_value.min_val, 1) << "error for shape (" << N0, N1, N2, N3 << ") and iteration " << iterate_msg;
    ASSERT_EQ(min_max_value.max_val, 1) << "error for shape (" << N0, N1, N2, N3 << ") and iteration " << iterate_msg;
    ASSERT_EQ(sum_value, N0 * N1 * N2 * N3) << "error for shape (" << N0, N1, N2, N3 << ") and iteration " << iterate_msg;
```
similarly elsewhere.

Reply:
Are you sure the `"error for shape (" << N0, N1, N2, N3 << ") and iteration "` syntax is valid?

Reply:
It's not valid syntax. My point is that you can adjust append to `ASSERT_EQ` instead of printing to `stdout`.

Reply:
Right. I experimented a bit, but it started to be a lot of duplicated code.